File: hash-table/word-pattern.js
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
var wordPattern = function(pattern, s) {
    let d1 = {};
    let d2 = {}

    let sl = s.split(" ")
    let pl = pattern.split("")

    if (sl.length !== pl.length) {
        return false;
    }

    for (let i = 0; i < sl.length; i++) {
        if (sl[i] === "constructor") {
            sl[i] = "contructor_haha"
        }
        if (pl[i] in d1 === false && sl[i] in d2 === false) {
            d1[pl[i]] = sl[i];
            d2[sl[i]] = pl[i];
        } else {
            if (d1[pl[i]] === sl[i] && d2[sl[i]] === pl[i]) {
                continue;
            }
            return false;
        }
    }

    // console.log(sl, pl, d1, d2);
    return true;
};

File: hash-table/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        
        s_li = s.split(" ")
        if len(s_li) != len(pattern):
            return False

        word_patt = {}
        patt_word = {}

        for char, word in zip(pattern, s_li):
            if char in patt_word and patt_word[char] != word:
                return False
            if word in word_patt and word_patt[word] != char:
                return False

            patt_word[char] = word
            word_patt[word] = char

            
        return True